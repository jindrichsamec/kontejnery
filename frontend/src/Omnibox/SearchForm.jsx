import React, { Component } from 'react'
import PropTypes from 'prop-types'
import 'whatwg-fetch'
import { observable } from 'mobx'
import { observer } from 'mobx-react'
import { interval, convertIntervalToDates } from '../utils/DateInterval'
import SearchQuery from '../model/SearchQuery'
import Spinner from '../Spinner'
import styled from 'styled-components'

const TermSelection = styled.a`
  line-height: 40px;
  margin: 0 8px;
  display: block;
  cursor: pointer;
  font-size: 1.2em;
  color: #000;

  &:after {
    font: normal normal normal 14px/1 FontAwesome;
    font-smoothing: antialiased;
    content: "\f107";
    position: absolute;
    right: 10px;
    display: inline-block;
    line-height: 40px;
  }
`

const TermSelect = styled.div`
  max-height: ${({open}) => open ? '200px' : '0'};
  overflow: hidden;
  transition: max-height 0.4s;
`

const Menu = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0;
`
const MenuItem = styled.a`
  padding: 5px 10px;
  margin: 0;
  display: block;
  line-height: 40px;

  &:hover {
    color: #fff;
    background: rgb(87,203,225);
    cursor: pointer;
  }
`

export default observer(class SearchForm extends Component {

  static propTypes = {
    onSearch: PropTypes.func.isRequired
  }

  intervalLabels = {
    [interval.TODAY]: 'Dnes',
    [interval.TOMORROW]: 'Zítra',
    [interval.NEXT_SEVEN_DAYS]: 'Příštích 7 dní',
    [interval.THIS_MONTH]: 'Tento měsíc',
  }

  obState = observable({
    searching: false,
    dateInterval: interval.NEXT_SEVEN_DAYS,
    showMenu: false
  });

  componentDidMount = () => {
    this.doSearch(this.obState.dateInterval)
  }

  dateToString = (date) => {
    if (date === null) {
      return ''
    }
    return date.toISOString();
  }

  getListEndpointUrl = (since, till) => {
    return `${process.env.REACT_APP_API_HOST}/api/list?since=%SINCE%&till=%TILL%`
      .replace('%SINCE%', this.dateToString(since))
      .replace('%TILL%', this.dateToString(till));
  }

  handleClick = (value) => {
    this.obState.dateInterval = value
    this.obState.searching = true
    this.obState.showMenu = false
    this.doSearch(value)
  }

  toggleMenu = () => {
    this.obState.showMenu = !this.obState.showMenu
  }

  doSearch(interval) {
    const {since, till} = convertIntervalToDates(interval)
    SearchQuery.since = since;
    SearchQuery.till = till;
    const url = this.getListEndpointUrl(since, till)
    fetch(url).then(this.handleSuccess, this.handleFail)
  }

  handleSuccess = (response) => {
    this.obState.searching = false

    response.json().then(this.normalizeData)
      .then(data => this.props.onSearch(data));
  }

  normalizeData(json) {
    const data = json.data.map((item) => {
      const till = new Date(item.till);
      return {...item, till}
    });
    const slugs = []

    return data.filter((item) => {
      const included = (slugs.indexOf(item.slug) > -1)
      slugs.push(item.slug)
      return !included;
    });
  }

  handleFail = (response) => {
    this.obState.searching = false
    console.error('Error during request');
  }

  getLabel(interval) {
    return this.intervalLabels[interval];
  }

  renderMenu =(value, index) => {
    return <MenuItem key={index} onClick={() => this.handleClick(value)}>
      {this.getLabel(value)}
    </MenuItem>
  }

  render() {
    return(
      <div>
        <TermSelection onClick={this.toggleMenu}>
          {this.obState.showMenu ? 'Hledat kontejner na...' : this.getLabel(this.obState.dateInterval)}
          {this.obState.searching && <span style={{marginLeft:'10px'}}><Spinner /></span>}
        </TermSelection>
        <TermSelect open={this.obState.showMenu}>
          <Menu title={this.getLabel(this.obState.dateInterval)} id="search-form">
            {Object.keys(this.intervalLabels).map(this.renderMenu)}
          </Menu>
        </TermSelect>
      </div>
    )
  }

})
