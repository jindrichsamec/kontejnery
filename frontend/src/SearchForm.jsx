import React from 'react'

import { FormGroup, FormControl, Button } from 'react-bootstrap'

const SearchForm = (props) => {
  return(
    <div className="card card-block controller" >
      <FormGroup>
        <FormControl componentClass="select" placeholder="Select one...">
          <option value="today">Dnes</option>
          <option value="tomorrow">Zítra</option>
          <option value="thisweek">Tento týden</option>
          <option value="nextweek">Příští týden</option>
          <option value="all">Vše</option>
        </FormControl>
        <Button type="button" bsStyle="primary">Search</Button>
        <Button type="button" bsStyle="success">Locate</Button>
      </FormGroup>
    </div>
  )
}

export default SearchForm
