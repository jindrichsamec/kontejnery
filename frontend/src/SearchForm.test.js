import SearchForm from './SearchForm'

describe('SearchForm', () =>Â {

  it('renders without crashing', () => {
    const wrapper = shallow(<SearchForm onSearch={jest.fn}/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('normalize data from `/list` endpoint', () => {

    const json = {
      data: [
        {
          "coordinates": {
            "lat": 50.1299980318329,
            "lng": 14.4796317815781
          },
          "id": 54,
          "name": "T\u0159ebenick\u00e1 x Sebuz\u00ednsk\u00e1 (parkovi\u0161t\u011b) (Kobylisy)",
          "till": "Tue, 11 Apr 2017 19:00:00 GMT"
        },
        {
          "coordinates": {
            "lat": 50.1381542860827,
            "lng": 14.434243440628
          },
          "id": 57,
          "name": "Fo\u0159tova x Do \u00dadol\u00ed (\u010cimice)",
          "till": new Date("Thu, 13 Apr 2017 19:00:00 GMT")
        }
      ]
    };

    const expected = [
      {...json.data[0], till: new Date("Tue, 11 Apr 2017 19:00:00 GMT")},
      {...json.data[1], till: new Date("Thu, 13 Apr 2017 19:00:00 GMT")},
    ];

    const component = new SearchForm()
    expect(component.normalizeData(json)).toMatchObject(expected)

  });

})

