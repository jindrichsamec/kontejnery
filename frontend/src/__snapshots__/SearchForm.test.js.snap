exports[`SearchForm renders without crashing 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <FormGroup
    bsClass="form-group">
    <InputGroup
        bsClass="input-group">
        <FormControl
            bsClass="form-control"
            componentClass="select"
            defaultValue="THIS_WEEK"
            inputRef={[Function]}
            onChange={[Function]}
            placeholder="Select one...">
            <option
                value="TODAY">
                Dnes
            </option>
            <option
                value="TOMORROW">
                Zítra
            </option>
            <option
                value="THIS_WEEK">
                Tento týden
            </option>
            <option
                value="NEXT_WEEK">
                Příští týden
            </option>
            <option
                value="THIS_MONTH">
                Tento měsíc
            </option>
        </FormControl>
        <InputGroupButton
            bsClass="input-group-btn">
            <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="default"
                disabled={false}
                onClick={[Function]}
                type="button">
                <Icon
                    name="search" />
            </Button>
        </InputGroupButton>
    </InputGroup>
</FormGroup>,
  "nodes": Array [
    <FormGroup
      bsClass="form-group">
      <InputGroup
            bsClass="input-group">
            <FormControl
                  bsClass="form-control"
                  componentClass="select"
                  defaultValue="THIS_WEEK"
                  inputRef={[Function]}
                  onChange={[Function]}
                  placeholder="Select one...">
                  <option
                        value="TODAY">
                        Dnes
                  </option>
                  <option
                        value="TOMORROW">
                        Zítra
                  </option>
                  <option
                        value="THIS_WEEK">
                        Tento týden
                  </option>
                  <option
                        value="NEXT_WEEK">
                        Příští týden
                  </option>
                  <option
                        value="THIS_MONTH">
                        Tento měsíc
                  </option>
            </FormControl>
            <InputGroupButton
                  bsClass="input-group-btn">
                  <Button
                        active={false}
                        block={false}
                        bsClass="btn"
                        bsStyle="default"
                        disabled={false}
                        onClick={[Function]}
                        type="button">
                        <Icon
                              name="search" />
                  </Button>
            </InputGroupButton>
      </InputGroup>
</FormGroup>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <SearchForm />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": SearchForm {
        "_dateToString": [Function],
        "_getListUrl": [Function],
        "_handleFail": [Function],
        "_handleFormChange": [Function],
        "_handleSuccess": [Function],
        "_reactInternalInstance": [Circular],
        "componentDidMount": [Function],
        "context": Object {},
        "intervalLabels": Object {
          "NEXT_WEEK": "Příští týden",
          "THIS_MONTH": "Tento měsíc",
          "THIS_WEEK": "Tento týden",
          "TODAY": "Dnes",
          "TOMORROW": "Zítra",
        },
        "obState": Object {
          "dateInterval": "THIS_WEEK",
          "searching": false,
        },
        "props": Object {},
        "refs": Object {},
        "render": [Function],
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <FormGroup
          bsClass="form-group">
          <InputGroup
                    bsClass="input-group">
                    <FormControl
                              bsClass="form-control"
                              componentClass="select"
                              defaultValue="THIS_WEEK"
                              inputRef={[Function]}
                              onChange={[Function]}
                              placeholder="Select one...">
                              <option
                                        value="TODAY">
                                        Dnes
                              </option>
                              <option
                                        value="TOMORROW">
                                        Zítra
                              </option>
                              <option
                                        value="THIS_WEEK">
                                        Tento týden
                              </option>
                              <option
                                        value="NEXT_WEEK">
                                        Příští týden
                              </option>
                              <option
                                        value="THIS_MONTH">
                                        Tento měsíc
                              </option>
                    </FormControl>
                    <InputGroupButton
                              bsClass="input-group-btn">
                              <Button
                                        active={false}
                                        block={false}
                                        bsClass="btn"
                                        bsStyle="default"
                                        disabled={false}
                                        onClick={[Function]}
                                        type="button">
                                        <Icon
                                                  name="search" />
                              </Button>
                    </InputGroupButton>
          </InputGroup>
</FormGroup>,
        "_debugID": 2,
        "_renderedOutput": <FormGroup
          bsClass="form-group">
          <InputGroup
                    bsClass="input-group">
                    <FormControl
                              bsClass="form-control"
                              componentClass="select"
                              defaultValue="THIS_WEEK"
                              inputRef={[Function]}
                              onChange={[Function]}
                              placeholder="Select one...">
                              <option
                                        value="TODAY">
                                        Dnes
                              </option>
                              <option
                                        value="TOMORROW">
                                        Zítra
                              </option>
                              <option
                                        value="THIS_WEEK">
                                        Tento týden
                              </option>
                              <option
                                        value="NEXT_WEEK">
                                        Příští týden
                              </option>
                              <option
                                        value="THIS_MONTH">
                                        Tento měsíc
                              </option>
                    </FormControl>
                    <InputGroupButton
                              bsClass="input-group-btn">
                              <Button
                                        active={false}
                                        block={false}
                                        bsClass="btn"
                                        bsStyle="default"
                                        disabled={false}
                                        onClick={[Function]}
                                        type="button">
                                        <Icon
                                                  name="search" />
                              </Button>
                    </InputGroupButton>
          </InputGroup>
</FormGroup>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <SearchForm />,
}
`;
